### ISSUE TITLE
[AWS][1.3.0.beta0] node state is erroneously `NetworkUnavailable`

### ISSUE THREAD

**Issue by simonswine** [kubernetes/kubernetes#27347]
I am currently trying `1.3.0-beta.0` on AWS. Everything works fine, but all worker kubelets are not schedulable, because of `NoRouteCreated`:

```
# kubectl get nodes -o yaml 
[...]
    - lastHeartbeatTime: null
      lastTransitionTime: 2016-06-14T08:01:20Z
      message: Node created without a route
      reason: NoRouteCreated
      status: "True"
      type: NetworkUnavailable
[...]
```

Node events:

```
  FirstSeen     LastSeen        Count   From                                                            SubobjectPath   Type            Reason                  Message
  ---------     --------        -----   ----                                                            -------------   --------        ------                  -------
  9m            9m              1       {kubelet ip-172-20-130-39.eu-west-1.compute.internal}                           Normal          Starting                Starting kubelet.
  9m            9m              1       {kubelet ip-172-20-130-39.eu-west-1.compute.internal}                           Normal          NodeHasSufficientDisk   Node ip-172-20-130-39.eu-west-1.compute.internal status is now: NodeHasSufficientDisk
  9m            9m              1       {kubelet ip-172-20-130-39.eu-west-1.compute.internal}                           Normal          NodeHasSufficientMemory Node ip-172-20-130-39.eu-west-1.compute.internal status is now: NodeHasSufficientMemory
  9m            9m              1       {controllermanager }                                                            Normal          RegisteredNode          Node ip-172-20-130-39.eu-west-1.compute.internal event: Registered Node ip-172-20-130-39.eu-west-1.compute.internal in NodeController
  8m            8m              1       {kubelet ip-172-20-130-39.eu-west-1.compute.internal}                           Normal          NodeReady               Node ip-172-20-130-39.eu-west-1.compute.internal status is now: NodeReady
  8m            8m              1       {kube-proxy ip-172-20-130-39.eu-west-1.compute.internal}                        Normal          Starting                Starting kube-proxy.

```

Though `kubectl get nodes` output makes you think, everything is ok:

```
core@ip-172-20-128-156 ~ $ kubectl get nodes
NAME                                           STATUS                     AGE
ip-172-20-128-156.eu-west-1.compute.internal   Ready,SchedulingDisabled   4m
ip-172-20-130-39.eu-west-1.compute.internal    Ready                      2m
ip-172-20-130-40.eu-west-1.compute.internal    Ready                      2m
```

No pod can be scheduled:
```
I0614 08:54:47.986353       1 event.go:216] Event(api.ObjectReference{Kind:"Pod", Namespace:"kube-system", Name:"kubernetes-dashboard-4126893489-lkymq", UID:"00b35fe4-320d-11e6-adba-0ae6bf41bde9", APIVersion:"v1", ResourceVersion:"136", FieldPath:""}): type: 'Warning' reason: 'FailedScheduling' no nodes available to schedule pods
```

I suspect it's maybe related to this new options for the controller-manager:

```
--allocate-node-cidrs[=false]: Should CIDRs for Pods be allocated and set on the cloud provider.
--configure-cloud-routes[=true]: Should CIDRs allocated by allocate-node-cidrs be configured on the cloud provider.
```

On my vagrant version of the same cluster (same coreos version, almost the same config), 1.3.0.beta0 works fine.

A few more bullet points about the setup:

- 1 master behind ELB
- 2 worker 
- All on coreos 1010.5
- cloud-provider="aws" on kubelets, apiserver, controller-manager
- flannel vxlan
- same config works fine on kubernetes version 1.2.4
- everything runs via kubelet (cf. https://coreos.com/kubernetes/docs/latest/kubernetes-on-aws.html)

- [ ] Better docs/elease notes about this NetworkUnavailable check
- [ ] Better `get nodes` output, a node should show up as NetworkUnavailable instead of ready
- [ ] Log debugging information to the events of the node 

How to debug further? I assume kubelet sets this status `"NoRouteCreated", Message:            "Node created without a route"`, but it's not overriden by the controller-manager later.

**Comment by wojtek-t** [comment]
Route-controller is supposed to create routes:
https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/route/routecontroller.go#L125
If Route is not created for a node, the node is not available for scheduling.

If Route is not supposed to be created in that configuration anytime, then probably this condition is not correct:
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L1031

**Comment by simonswine** [comment]
@wojtek-t thanks for pointing out the involved code.

I am not too sure about the whole thing, but I think we only want to enable this if I am not running an overlay network myself. Do you have a clue how to disable the routes feature? Sure `--experimental-flannel-overlay` would disable it, but I control flannel with systemd.

- cloud provider get's called [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L1118)
  - it's calling a cloud provider if it's one is set
  - it's false for ` --experimental-flannel-overlay`

- aws returns always true: https://github.com/kubernetes/kubernetes/blob/release-1.3/pkg/cloudprovider/providers/aws/aws.go#L751

**Comment by wojtek-t** [comment]
>
I am not too sure about the whole thing, but I think we only want to enable this if I am not running an overlay network myself.

Yes - that is correct.

>
Do you have a clue how to disable the routes feature? Sure --experimental-flannel-overlay would disable it, but I control flannel with systemd.

I don't know - this may require some code changes.

**Comment by simonswine** [comment]
@justinsb: maybe you can help on this?

**Comment by justinsb** [comment]
Triaging as 1.3, P1.  My understanding is that it blocks all operation of k8s when used with flannel (or any networking provider other than the built-in ones?)  Verifying further.

My guess is that we just want to disable this check when `configure-cloud-routes=false` or when `experimental-flannel-overlay=false`

**Comment by justinsb** [comment]
Ah, so kubelet registers the node with routing status "broken", and it is the responsibility of the route controller to configure the node routing status to "healthy" once it has configured it.  That's nice, because it means we won't schedule pods to a node until the routing controller has configured it.  It also is a scalable way to add more required controllers.

In this case, that means flannel must do it.  So, probably a P2 or docs issue for k8s, but probably a P0 for flannel.

I agree with the initial checklist.  @simonswine can you explain the issue with passing `experimental-flannel-overlay` while using systemd - I'm not familiar with it.

We should let the networking providers know that they must do this for 1.3.  I'm not sure whether somehow flannel is different in this regard.  cc @kubernetes/sig-network

**Comment by simonswine** [comment]
Thanks @justinsb for your support with that.

I was not aware of this `--experimental-flannel-overlay` flag before I hit that issue and I had to dig through the code.

It seems to me with enabling it, I have the exact same behaviour like I had before with versions 1.2.x, this should be part of upgrade notes. I am not too sure about how stable the route controller is. Maybe the default value for `--experimental-flannel-overlay` should be `true` for 1.3?

**Comment by goltermann** [comment]
Is this needed for 1.3?  It sounds like it may be.  If so, please move to P1.  Otherwise, remove the 1.3 milestone.

**Comment by davidopp** [comment]
As @justinsb described above, third-party network providers that have not been updated to be "NetworkUnavailable"-aware are going to break.

The simplest fix is to disable the new behavior for such providers. Even if there's a way to make it work for flannel, other providers are going to be broken also.

**Comment by davidopp** [comment]
@erictune This isn't AWS-related. I see you just assigned Justin but I'm going to unassign him. It needs an owner.

**Comment by simonswine** [comment]
@davidopp I agree that this a more generic thing, I am not too sure how the other cloud-providers respond to `Routes()`

I see two possible simple solutions for the whole issue:

AFAIK in the case where the network provider is not supporting `NetworkUnavailable` the controller manager (`configure-cloud-routes=false`) should be aware of that and it could just mark all the NetworkUnavailable states in the nodes as 'resovled' without doing anything else.(don't setup routes, do not check anything)

The other solution is to do it on the kubelets by making `--experimental-flannel-overlay=true` the default and maybe renaming it too be more generic argument and without `experimental`, because IMHO the experimental stuff is happening when you disable that option.

**Comment by erictune** [comment]
@wojtek-t I think you added `NetworkUnavailable`.  Is that right?  It seems like the current behavior is going to break networks that worked with 1.2.  So, this is arguably a release blocking regression.  I think it needs to somehow know that the network provider is not one of the once that knows how to say when it is ready, like by adding something like `CanReportReadiness()` to the network provider API.  Can you do that or suggest someone else?  Would like to get all 1.3 blockers fixed by friday

**Comment by wojtek-t** [comment]
@erictune - yes that was me, sorry for breaking;
However, to be honest I think that what you are suggesting is not going to fix the problem (or I misunderstood your proposal - then please correct me).
Basically the problem in this case is having an overlay network. If it was used based on AWS networking it would work, because AWS supports routes interface and thus routecontroller would create them and set those conditions correctly.

So my understanding is that in this case, we simply switch off "routes interface" by setting "--configure-cloud-routes" option in controller-manager - am I right? @justinsb  @simonswine - can you please confirm?

If that's the case, Kubelet should never create "NetworkUnavailable" condition, thus the cleanest way to solve it is to simply add another flag to kubelet that would gate setting NetworkUnavailbale condition.

**Comment by davidopp** [comment]
@wojtek-t what part of what @erictune wrote are you disagreeing with? I assume just the "like by adding something like CanReportReadiness() to the network provider API" part? His comment seemed generally accurate to me.

In any event, I think the bottom line is that any approach that causes Kubelet to set `NetworkUnavailable = false`, or not set `NetworkUnavailable`, for network providers that aren't "aware" of the NetworkUnavailable condition, is fine.

**Comment by wojtek-t** [comment]
So I talked with @davidopp offline about this issue, and we think that the easiest fix for 1.3 is to not set "NetworkUnavailable" condition to fale on any cloud provider different than GCE/GKE. And fix it properly for other cloud providers in 1.4.

If we do that, this won't be regression, because there wasn't NetworkUnavailable condition in 1.2 at all.

Such PR would be super trivial so I will send it anyway - if you don't agree, we can simply close it.

**Comment by davidopp** [comment]
SGTM

BTW any misinterpretation @erictune had is my fault, as I had summarized the issue to him.

**Comment by justinsb** [comment]
I _think_ though that the problem will still exist for flannel with GCE?  (Not sure if that is a very common scenario - @philips  ?)

I agree that for 1.3 we likely need something like #27525.  I like the approach though of having the node come up broken when it needs extra work, and then having a controller mark each condition as fixed as it is configured.  The core problem is that kubelet doesn't know if an "aware" controller is present.

As an alternative, it would also be pretty easy to create a simple add-in controller that could blindly set this state to healthy, so then flannel/calico configurations would just have to drop that in as a daemonset or similar.  But I think we only need this if GCE + flannel (or some other networking) is important

**Comment by wojtek-t** [comment]
@justinsb - in GCE we have support for flannel in our configuration configs, e.g.:
https://github.com/kubernetes/kubernetes/blob/master/cluster/gce/config-default.sh#L137

If you use flannel by this path, this will trigger setting "--experimental-flannel-overlay":
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubelet/app/options/options.go#L262
and this would result in not setting "NetworkUnavailable" in kubelet:
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L1128

So unless you are using flannel in "your fancy way", it will work.

**Comment by rootfs** [comment]
@wojtek-t do we need to add `--experimental-flannel-overlay` to `hack/local-up-cluster.sh` to work on aws/gce?

**Comment by erictune** [comment]
@wojtek-t, please add a flag that disables the problem reporting, and does nothing else.

**Comment by davidopp** [comment]
Please make sure to also fix #26983  (it's referenced above but just wanted to call attention to it; I don't think #27525 fixes it).

**Comment by wojtek-t** [comment]
I asked for check in the other issue - my feeling is that it fixes the other bug too.

**Comment by erictune** [comment]
@davidopp Can you summarize status of this issue.

**Comment by wojtek-t** [comment]
@erictune - I think this is fixed (i.e. I don't know how to reproduce it). If that's not the case, I need some repro instructions.

**Comment by justinsb** [comment]
I propose closing for the same reason as https://github.com/kubernetes/kubernetes/issues/26983#issuecomment-227617059 .  We can always reopen if needed.

I opened https://github.com/kubernetes/kubernetes/issues/27828 to track re-enabling this functionality in 1.4 - it's probably not too hard, just not really practical in 1.3.0

**Comment by simonswine** [comment]
Just tested this with `1.3.0-beta.2` it is fixed there

**Comment by wojtek-t** [comment]
@simonswine - awesome. Thanks for confirmation!
